acme: use the hotplug system

hgl opened the pull request:
Signed-off-by: Glen Huang <heyhgl@gmail.com>

Maintainer: @tohojo
Compile tested: not yet
Run tested: not yet

cc httpd maintainers and major contributors: @micmac1 @Ansuel @neheb @peter-stadler @jow- @feckert @nbd168

## Why

This PR might come off as presumptuous, since it removes a lot of code and is not backward compatible, but allow me to explain myself.

I think the acme package suffers from assuming too much responsibilities for interacting with httpds. It currently does the following:

1. Finding and turning off httpds listening on port 80 when standalone mode is specified.
2. Making httpds reload once certs are issued/renewed.

The way it does them is pretty brittle (also dangerous, since it modifies other package's config files), and it has to account for every supported httpd, which is unmaintainable.

The situation can be improved, if:

1. This package uses the hotplug system to send issue/renew events
2. Each httpd package plugs into the system, and coordinates with this package via such events.

## How

This package has to interact with httpds in the following ways:

1. HTTP-01 via webfoot
2. HTTP-01 via acme.sh standalone mode
3. Make the httpd reload the new certificates

For `#1`, currently the user is allowed to specify a path. With the hotplug system however, it should be a well-defined path like `/var/run/acme/challenge`, so each httpd can just serve challenge from this dir to make acme work.

For `#2`, it should be the user's responsibility to not use the standalone mode when a httpd is listening on 80. The way this package currently finds httpds is flawed, since it checks if port 80 is used on any interfaces, but standalone mode only requires 80 on WAN. Turning them off and restarting them later is also brittle.

For `#3`, if the hotplug system is employed, this package will call hotplud.d/acme, and scripts in it will be called with environment variables denoting if a domain is (about to be issued/issued/renewed/failed to do the proceedings) along with others that are [set by acme.sh](https://github.com/acmesh-official/acme.sh/blob/4c32bc8e22075d3333f49973016a62983205f0aa/acme.sh#L3518). Each httpd should know better how it should reload itself and add certificates to its config when first issuing.

This package also supports user script options for setting up and cleaning up. They can be removed as the user can just put scripts in the hotplug system.

## Implementation

This PR has not been tested. It only serves to start the conversation, and demonstrates how it should work in theory. Once the design is finalized, I will make sure it's actually runnable and hopefully bug-free. At the moment, I'd be interested to know if @tohojo and httpd maintainers and major contributors think this approach is sound, and if adding acme support in each httpd package is something the maintainers and contributors would oppose to.

To reload a httpd after successfully issuing or renewing a certificate, a hotplug script could look something like:

```shell
case "$ACTION" in
issued)
	# change config so it specifies the certificates

	/etc/init.d/nginx reload
	;;
renewed)
	/etc/init.d/nginx reload
	;;
*) exit ;;
esac
```

This script ideally should ideally live in a separate package like nginx-acme, so it can be optionally enabled.

An even simpler approach is to require that the user runs this package at least once (either via the command line or LuCI), before specifying the certificates in the httpd config. That way the hotplug script no longer needs to mess with user's config.

I also wonder, since it's not backward compatible, should it be implemented as a separate package? Or it should be updated in place only for a new OpenWrt major version?

## Details

I also took the liberty to change some small things, let me know if you think any of them are terrible ideas:

1. It should not use a init script since it's not a daemon.
2. Cron job is added once by ipk, instead of checking repeatedly by command.
3. Cron job is run each month, instead of every day.
2. `cert` section is enabled by default, to follow the convention.
3. Renamed option `use_staging` to `staging`.
4. Use `ec-256` as `keylength` by default.
5. No need to resend INT signal, since it's not prevented.
6. Rename `err` to `logerr`
8. Cleaning up is automatically run at shell exit, not at the end of each issue_cert.
9. Keep issuing certificates if one fails.
10. Not printing logs where they should be clear either from acme.sh's logs or passed flags

## Questions

I'm not sure what `CURL_CA_BUNDLE` and `NO_TIMESTAMP` are for?

peter-stadler commented:
I think this could be a good concept, although I do not understand it fully (as I do not know the details neither of acme nor the hotplug system).

It could be viable for *Nginx*, there is one thing though:
How would a (Nginx) hotplug script get the information, for which domains the configuration should be updated (and where the corresponding certificates are located)?

In the current version the acme.sh script is calling:
```
for domain in $domains; do
	nginx-util add_ssl "${domain}" acme "${domain_dir}/fullchain.cer" \
		"${domain_dir}/${main_domain}.key"
```
I do not hesitate to adopt `nginx-util`, but do not have much time (I have still the open issue with wildcard certificates https://github.com/openwrt/packages/issues/16909, but maybe that can be handled in a good way then, too.)

Ansuel commented:
@peter-stadler i assume we should add a script for the hotplug.d. When the event is fired, we can cal the nginx-util bin to update the certificate and then reload nginx... I really like this kind of system and would work really well... (Much better than using cron job and force a restart of all the connected services)

hgl commented:
@peter-stadler This packages uses acme.sh to do the actual ACME stuff, and acme.sh sets the `Le_Domain` environment variable to the main domain. The hotplug system should call a script multiple times for multiple (main) domains, so the snippet you posted should simply be turned into:

```shell 
nginx-util add_ssl "${Le_Domain}" acme "${domain_dir}/fullchain.cer" \
		"${domain_dir}/${Le_Domain}.key"
```

@peter-stadler @Ansuel  However, as I mentioned in the OP. I do think modifying a user's config in a hotplug script might be brittle, since just adding certificate options can failed to work if the user haven't added other required ssl options.

Asking users to run acme at lease once, and then add the ssl related config themselves is another option. This also can make the hotplug script simpler.

hgl commented:
> I'd like to keep the switch backwards-compatible at a config level

That's reasonable. Most obsolete options should simply be ignored and renamed ones should be able to fallback to old ones.

> I think we should wrap the hotplug execution and redefine the
  environment variables to something well-defined

Good point.

> There was a stalled-out effort to make uacme interchangeable with acme.sh

I see the uacme package mostly duplicates the `run-acme` command with this package. I wonder if this package should mimic ddns, so it employs a plugin system where users can choose which ACME client to use?

>  I wouldn't be opposed to swapping the defaults either

Do you mean use uacme by default? I prefer its small footprint too, but last time I checked, DNS01 challenge is not supported out of the box. Users have to provide scripts themselves for each DNS vendor if I'm not wrong. 

> I fear it will become unmanageable to keep things in sync across packages if they are separated

That shouldn't be the case if each package's responsibility is clearly defined. For this package, I'd like it to be "issue or renew the certs and tell the world after it's done", for other packages, they can choose what do to when being told certs are issued or renewed: reload a httpd, or upload the cert somewhere etc. It'd be unmanageable if this package has to cater for every use case of issued/renewed certs.

> If we have another way of doing that we don't need the init script.

Maybe @jow- has a better idea?

> That is probably too seldom

Agreed, I forgot busybox crontab is unable to run jobs retroactively.

> That's awfully close to 'logger'

Maybe `log info` and `log err`?

> That's not really compatible with users-defined cleanup scripts

By "cleaning up" I mean cleaning up required by `run-acme`, in this case the removing of iptables rules for webroot/standalone mode.

Users-defined cleanup scripts are no longer supported, since the hotplug script will be called when a cert is failed to issue/renew, they should just put such scripts in hotplug.

> so at least a summary of the config should be printed at the start of the run

Do you mean the options that are used by the current run? I thought printing final acme.sh cmd (along with the flags we passed) was enough to debug and tell the user what the options eventually converted to.

> CURL_CA_BUNDLE does what it says on the tin...

Thanks for the heads up.

tohojo commented:
Glen Huang ***@***.***> writes:

>> There was a stalled-out effort to make uacme interchangeable with acme.sh
>
> I see the uacme package mostly duplicates the `run-acme` command with
> this package. I wonder if this package should mimic ddns, so it
> employs a plugin system where users can choose which ACME client to
> use?

So what we were discussing at the time (IIRC) was to have the run script
and config be common between the two packages (in an 'acme-common'
package they could both depend on).

With the architecture you're proposing I guess that would actually be
easier; have 'acme-common' contain the config parsing and the hotplug
infrastructure and then have each of 'acme' and 'uacme' just install the
binary and a script providing only the 'run' bit for that particular
binary, and the hotplug wrapper that translates the runner hook envvars
to the common API we agree on. I guess you could call that a 'plugin
system' but let's keep it as simple as possible :)

(I'm not familiar with how ddns works, so if the above fits that, great!)

>>  I wouldn't be opposed to swapping the defaults either
>
> Do you mean use uacme by default? I prefer its small footprint too,
> but last time I checked, DNS01 challenge is not supported out of the
> box. Users have to provide scripts themselves for each DNS vendor if
> I'm not wrong.

Ah, right, I thought they were basically at feature parity, but I guess
some people will want the DNS provider scripts from acme.sh; okay, so
let's keep that as the default but support uacme as well as outlined
above. 

>> I fear it will become unmanageable to keep things in sync across packages if they are separated
>
> That shouldn't be the case if each package's responsibility is clearly
> defined. For this package, I'd like it to be "issue or renew the certs
> and tell the world after it's done", for other packages, they can
> choose what do to when being told certs are issued or renewed: reload
> a httpd, or upload the cert somewhere etc. It'd be unmanageable if
> this package has to cater for every use case of issued/renewed certs.

Right, sure, let's split the responsibilities so there's a separate hook
script for each daemon; the question is more if we should package those
scripts as part of the 'acme' package, instead of offloading them on the
daemon maintainers by putting them in the uhttpd, nginx etc packages.
I'm mostly concerned about someone updating their acme package and
suddenly they have no hook scripts :)

>> That's awfully close to 'logger'
>
> Maybe `log info` and `log err`?

Sure, that's fine.

>> That's not really compatible with users-defined cleanup scripts
>
> By "cleaning up" I mean cleaning up required by `run-acme`, in this
> case the removing of iptables rules for webroot/standalone mode.
>
> Users-defined cleanup scripts are no longer supported, since the
> hotplug script will be called when a cert is failed to issue/renew,
> they should just put such scripts in hotplug.

But the 'failure' script will not be run (by the ACME runner) if the
renewal is interrupted by SIGINT, so we need to handle that case as
well...

>> so at least a summary of the config should be printed at the start of the run
>
> Do you mean the options that are used by the current run? I thought
> printing final acme.sh cmd (along with the flags we passed) was enough
> to debug and tell the user what the options eventually converted to.

It's nice to have a "these are the options I got from uci" as well. Can
just be a simple message at the start, like "Running acme certificate
runner in standalone mode for domain xxx.yy with options A/B/C" so that
it's easy to spot if a failure is due to a configuration error or a bug
in run-acme...

hgl commented:
> just install the binary and a script providing only the 'run' bit for that particular
binary, and the hotplug wrapper that translates the runner hook envvars
to the common API we agree on

This is what I have in mind too.

> so let's keep that as the default but support uacme as well as outlined
above

What do you think of this design, which is inspired by ddns:

1. All acme clients should live in this package, which use one Makefile to build multiple sub-packages for each of them. This ensures code reuse and every client following the convention required by the "common" package.
2. Each client sub-package is named as `acme-something`. For `acme.sh`, it could be `acme-acmesh`. The `acme-` prefix is a signal that `luci-app-acme` should work with any of them.
3. Each client sub-package should either conflict with each other, or the config should allow users to specify the client to use. The latter can make the run script much more complex, I'm not sure if it's worthwhile.
3. The common package is named `acme`. If no real client package is installed, executing its bin should exit with an error saying things like a client package is missing.

> the question is more if we should package those scripts as part of the 'acme' package

I see your point. I'm worried this might encourage feature bloat. If some package needs to upload the certs to the cloud after being renewed, it might mimic the provided httpd script and submit a PR to add such uploading script to this package.

> I'm mostly concerned about someone updating their acme package and
suddenly they have no hook scripts :)

It should be fine if they also update the httpd package.

> But the 'failure' script will not be run (by the ACME runner) if the renewal is interrupted by SIGINT

Ah, I see. You mean the case where hotplug script is called before the cert is issued/renewed, but the run script is interrupted during the issuing/renewing? Yeah, it should be handled. Maybe trap INT and call the hotplug script with error (treating it the same as issue/renew error)?

> It's nice to have a "these are the options I got from uci" as well
 
That's reasonable.

tohojo commented:
Glen Huang ***@***.***> writes:

> What do you think of this design, which is inspired by ddns:
>
> 1. All acme clients should live in this package, which use one
> Makefile to build multiple sub-packages for each of them. This ensures
> code reuse and every client following the convention required by the
> "common" package.
> 2. Each client sub-package is named as `acme-something`. For
> `acme.sh`, it could be `acme-acmesh`. The `acme-` prefix is a signal
> that `luci-app-acme` should work with any of them.
> 3. Each client sub-package should either conflict with each other, or
> the config should allow users to specify the client to use. The latter
> can make the run script much more complex, I'm not sure if it's
> worthwhile.
> 3. The common package is named `acme`. If no real client package is
> installed, executing its bin should exit with an error saying things
> like a client package is missing.

SGTM. I think it's fine to have the packages conflict, at least to begin
with. If someone shows up with a compelling use case for needing
multiple clients simultaneously we can always reconsider :)

>> the question is more if we should package those scripts as part of the 'acme' package
>
> I see your point. I'm worried this might encourage feature bloat. If
> some package needs to upload the certs to the cloud after being
> renewed, it might mimic the provided httpd script and submit a PR to
> add such uploading script to this package.

Well, then we can say no ;)

But okay, I do agree that separation of concerns is nice, and even
though it'll add a little bloat to each daemon package (installing the
hotplug script regardless of whether acme is used), maybe that is OK. I
guess we can try it out in a common PR and see if we get any pushback
from any of the daemon maintainers...

>> I'm mostly concerned about someone updating their acme package and
> suddenly they have no hook scripts :)
>
> It should be fine if they also update the httpd package.

Yeah, 'if'. But okay, maybe it's a theoretical concern; we should make
sure to merge things together, though (i.e., adding hotplug scripts to
the daemons in the same PR that changes how acme itself works).

>> But the 'failure' script will not be run (by the ACME runner) if the renewal is interrupted by SIGINT
>
> Ah, I see. You mean the case where hotplug script is called before the
> cert is issued/renewed, but the run script is interrupted during the
> issuing/renewing? Yeah, it should be handled. Maybe trap INT and call
> the hotplug script with error (treating it the same as issue/renew
> error)?

Yup, something like that would work.

hgl commented:
@tohojo I just implemented what we discussed. Also tested on a real OpenWrt machine. 

## Changes

1. The `acme` package installs `/use/bin/acme`.
2. It support cmds like `acme get` and `acme cert <domain>`
3. It calls `/usr/lib/acme/hook`, which should be installed by an ACME client package.
4. Since a package's makefile can only contain one `Build/Configure`, each ACME client package must live in its own folder.
5. Each ACME client package should provide an `acme-client` virtual package.
6. Each httpd ACME package should depend on `acme-client`.
7. Added `nginx-acme` as an example. Users are expected to run `acme get` at least once before adding ssl options to `nginx.conf`.

## Questions

1. To make each ACME client package conflicts with each other, it seems they have to know each other's names. I wonder if there is a way to specify "each package that provides `acme-client` should conflict with each other"?
2. Still need help on how to make LuCI trigger `acme get` on change.

hgl commented:
Any maintainer is available to review the PR?

It'd be sad to let it go stale.

tohojo commented:
It's on my list, but won't have time until next week...

tohojo reviewed (changes requested):
I think it's a good start, and I like the overall architecture. Thanks for working on this!

Still some details to work out, though, see the individual review comments.

As for the conflict, since we only have two implementations, I think it's fine to just have the conflicts spelled out; if opkg ever gains support for "only one provider for each meta package" then we can use that at that point. For getting UCI to call acme on update, I think that goes through the init system? So why not just keep the init script?

neheb commented:
I advise running against shellcheck -s dash. Applies to some of @tohojo 's comments.

hgl commented:
@tohojo Thanks for taking the time reviewing this. Will address your comments when time allows (hopefully next  week).

@neheb Good suggestion. I'll make sure to use it.

I had another thought will reading the comments: what about giving well-defined paths to issued certificates? For example, every acme-client should make sure that after a successful issuing/renewing, a certificate issued with example.com as the main domain, should be put at `/etc/ssl/example.com.crt`, and fullchain at `/etc/ssl/example.com.fullchain.crt`, and the private key at `/etc/ssl/private/example.com.key` for example.

This simplifies httpd configs and hotplug scripts.

tohojo commented:
We could make the runner scripts create symlinks? That's what the letsencrypt certbot does :)

Probably good to put them in a subdir to avoid name clashes, so /etc/ssl/acme ?

hgl commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,122 @@
+	log info "Running ACME for $main_domain"
+
+	if [ -e "$domain_dir" ]; then
+		if [ "$staging" = 0 ] && grep -q "acme-staging" "$domain_dir/$main_domain.conf"; then
```
Since we're just testing for equality, I think `=` is less typing and looks intuitive. I agree its usages are inconsistent, and they're now fixed.

hgl commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,122 @@
+		else
+			acme_args="--renew --home "$state_dir" -d "$main_domain" $acme_args"
+			log info "$ACME $acme_args"
+			trap 'export ACTION=renewed-failed;hotplug-call acme;exit 1' INT
```
> I think we should wrap the hotplug-call

Such script provides really thin abstraction, and invents our own version of `hotplug-call`, which can be confusing IMHO. 

> Also, please document somewhere which actions are supported

Good suggestion, I intend to add the doc as a comment block in /usr/bin/acme once the API design is fixed.

> the hotplug wrapper needs to translate the environment variables

What needs to be translated? Shouldn't each hotplug script just receive config options passed down from /usr/bin/acme?

tohojo replied:
So my point was that acme.sh sets $Le_Domain and $Le_Alt when running its hook scripts, while uacme will set something different. And hotplug scripts really shouldn't need to care about which backend is actually used, though. But OK, if hotplug scripts are only expected to read the variables exported by the wrapper itself, that'll work; this should be clearly documented, though :)

hgl replied:
Do you think the doc should live in OpenWrt wiki page or within the bin script?

hgl commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,122 @@
+	"$ACME" \
+		$acme_args \
+		--pre-hook 'export ACTION=prepare;hotplug-call acme' \
+		--post-hook 'export ACTION=issued;hotplug-call acme' \
```
Good catch, fixed.

hgl commented on net/acme/files/acme.sh:
```diff
@@ -0,0 +1,179 @@
+	[ "$enabled" -eq "1" ] || return
+
+	load_options "$section"
+	if [ -z "$dns" -a "$standalone" = 0 ]; then
```
Fixed.

hgl commented on net/acme/files/acme.sh:
```diff
@@ -0,0 +1,179 @@
+	if [ -z "$staging" ]; then
+		config_get_bool staging "$section" staging 0
+	fi
+	export staging
```
I'm trying to follow [ddns' conversion](https://github.com/openwrt/packages/blob/86b8b30683cb7e338324c4cba5bee69cc6f1dcd6/net/ddns-scripts/files/usr/lib/ddns/update_cloudflare_com_v4.sh#L25). Using same-case variables names eliminating a translating step and is pretty intuitive. I don't think there is a need to document, the variables are already documented in the acme config file.

As for the section name, I don't think we should export that. All config values are already exported, so a hotplug script already has all the values it needs.

hgl commented on net/acme/files/acme.sh:
```diff
@@ -0,0 +1,179 @@
+	set -- $domains
+	path="$($HOOK print-cert-path "$1")"
+	echo "$path"
+	openssl x509 -in "$path" -text -noout
```
Good point, removed.

hgl commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,122 @@
+# Needed by acme.sh
+export CURL_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
+export NO_TIMESTAMP=1
+
```
Good pointed, added.

hgl commented on net/nginx/files/acme.hotplug:
```diff
@@ -0,0 +1 @@
+[ renewed = "$ACTION" ] && /etc/init.d/nginx reload
```
As I mentioned in the OP, I think messing with user's config is a mistake, it's dangerous and flaky. Requiring the user to run /usr/bin/acme at least once before specifying ssl in a httpd solves all the problems, and is something reasonable to require IMHO.

tohojo replied:
Right, but there's also a backwards compatibility issue here (see my other comment below). Now, arguably if you're upgrading and have already issued a certificate, your nginx config should already contain the cert path, so as long as that stays stable things should continue to work.

However, we should at least detect if the different `update_*` variables are set in the config, and emit a deprecation warning (maybe combined with a grep for `$domain_dir` in the config to get an indication as to whether things will continue to work?)

hgl replied:
I see your point. Although with the new design, a httpd is supposed to be automatically updated, and the `update_*` options are simply ignored. I wonder if it's necessary to point this out in a warning log, since nothing should be broken and continue to work?

tohojo replied:
Right, plain nginx will continue to work on an upgrade as long as `$domain_dir` stays the same, which it does. However, haproxy won't because that relies on a different name, and uhttpd won't because that doesn't have a reload hook anymore. Also, nginx-util is no longer called, so anyone using that will break.

hgl commented on net/acme/files/acme.sh:
```diff
@@ -0,0 +1,179 @@
+	export acme_server
+	config_get days "$section" days
+	export days
+	config_get standalone "$section" standalone 0
```
There is nothing to be compatible with. `webroot` is now a predefined path, so the old path is not used anymore.

tohojo replied:
Exactly. So if someone is upgrading with an old config that sets a custom webhook, certificate renewal will now fail. We can change the default and encourage using the new functionality (so a deprecation warning is fine), but no straight-up breaking old configs.

hgl replied:
Maybe there is some misunderstanding? `webroot` is where a httpd gets the challenge token, which is generated at runtime. Ignoring a previously custom `webroot` shouldn't break renewal, since the new standardized location will work.

hgl commented on net/acme/files/acme.sh:
```diff
@@ -0,0 +1,179 @@
+fi
+
+case "$1" in
+get)
```
That's a nice feature to have, but not trivial to implement, and it's only usable in CLI. Maybe leave that to a future version?

hgl commented:
@tohojo life gets in the way. Didn't have time to work on this until now. Sorry about the delay. It's good for review.

neheb commented:
What's the status here? ping @tohojo

hgl commented:
I have updated to solve the conflict with the master branch.

I also added the code that creates a symbolic link to `/etc/ssl/acme/$main_domain.chain.cer` to support `ssl_trusted_certificate` in nginx.

Would appreciate it if any maintainer could offer to review or merge.

tohojo commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,128 @@
+		ln -s "$domain_dir/$main_domain.cer" /etc/ssl/acme
+		ln -s "$domain_dir/$main_domain.key" /etc/ssl/acme
+		ln -s "$domain_dir/fullchain.cer" "/etc/ssl/acme/$main_domain.fullchain.cer"
+		ln -s "$domain_dir/ca.cer" "/etc/ssl/acme/$main_domain.chain.cer"
```
While we're standardising locations, should we also add a 'combined' version (cat'ing .key and .fullchain.cer and linking it as `/etc/ssl/acme/$main_domain.combined.cer`)?

hgl replied:
I think it'd be good idea to stick to what certbot offers. If an httpd requires a file like that, it can create one itself in the hook.

tohojo reviewed (changes requested):
OK, so I think we're mostly looking good on the code side, with a few nits below.

However, the backwards compatibility story still needs some work. Specifically:

- We'll need hotplug scripts for uhttpd, haproxy and nginx-util
- We may need to shuffle around the package name and dependencies a bit. Right now you're making `acme-acmesh` depend on `acme`, which means that someone who has only installed `acme` will end up with a partial install on an upgrade.

On way to resolve that latter case could be to name the base package something like `acme-common` and then make a new `acme` package that depends on `acme-acmesh`. But if you have a better solution for this I'm all ears :)

hgl commented:
> We'll need hotplug scripts for uhttpd, haproxy and nginx-util

I think it'd be a good opportunity to let httpd maintainers try to contribute such hotplug scripts. It provides the opportunity to both try out the new architecture and hopefully elicit some feedbacks. Thoughts @gladiac @Ansuel @peter-stadler @jow- ?

I'm not sure if nginx-util should itself provide a hotplug script. It's only a CLI that adds ssl directives to nginx.conf. And in this package it's only previously used for new certificates, so there shouldn't be compatibility issues if I'm not missing anything.

> We may need to shuffle around the package name and dependencies a bit

What about we keep the old acme package, name the package in this PR `acme-common`, and make `acme-acmesh` depends on it. So to use the new architecture, users have to remove `acme` and install a different package.

hgl commented on net/acme-acmesh/files/hook.sh:
```diff
@@ -0,0 +1,128 @@
+
+case $cmd in
+get)
+	set -- "$ACME"
```
updated

tohojo replied:
Missed one `"$@"` invocation :)

hgl replied:
Sorry, not sure I follow. You mean the first `set` invocation should contain `"$@"`? Since `acme get` current doesn't support any flags, I think we should be defensive here and ignore all flags. The CLI can always be changed latter.

tohojo replied:
No, I meant you missed translating one instance of `"$@"` to `"$ACME" "$@"` (on line 46) :)

hgl replied:
Ah, nice catch. Fixed.

hgl commented:
I also added a procd script to handle UCI changes. 

@tohojo I wonder what `check_cron` was for in the procd script's boot callback, previously? Shouldn't cron setting be persistent?

hgl commented:
I have:

1. Renamed the package to `acme-common`.
2. Added the `acme` meta package.
3. Added acme hooks for haproxy. uhttpd lives in another repo, so PR will be sent later to that repo.
4. Used UCI defaults to address the cron issue. So no need for the boot hack.

tohojo commented on net/haproxy/files/acme.hotplug:
```diff
@@ -0,0 +1 @@
+[ renewed = "$ACTION" ] && /etc/init.d/haproxy reload
```
This needs to also combine the key and fullchain cert into a single .pem file...

hgl replied:
Fixed.

tohojo reviewed (changes requested):
One missing bit in the haproxy hook file. @lialosiu care to test if everything still works after this PR?

Also, I'm wondering if it's a bit overkill to pull out the hotplug scripts into separate packages? Sure, they're technically a separate feature, but the hook files are also one-line shell scripts, so we could just always include them in the base packages?

hgl commented:
> Also, I'm wondering if it's a bit overkill to pull out the hotplug scripts into separate packages?

Agreed, separate *-acme packages are now removed. 

Looks like either other maintainer are too busy to chime in here, or this PR has too many comments that nobody bothers to read anymore.

If you no have no further objections, @tohojo. How about just merge this PR and go from there? I will try to be active for the next fews days and address any bugs that might come up. This PR is in appending state for too long, I'm worried that its own weight starts to enervate instead of encouraging contributions.

Thanks for the (so many) careful reviews so far, @tohojo!

tohojo commented on net/nginx/Makefile:
```diff
@@ -372,6 +372,9 @@ ifeq ($(CONFIG_NGINX_NAXSI),y)
 	$(if $(CONFIG_NGINX_NAXSI),$(chmod 0640 $(1)/etc/nginx/naxsi_core.rules))
+
+	$(INSTALL_DIR) $(1)/etc/hotplug.d/acme
+	$(INSTALL_DATA) ./files/acme.hotplug $(1)/etc/hotplug.d/acme/00-nginx
```
The nginx Makefile is not using `$(AUTORELEASE)`, so you'll need to bump the package revision here as well...

hgl replied:
Good catch. Done.

tohojo reviewed (changes requested):
Yes, I agree; I think this is in pretty good shape now, let's merge it and deal with any fallout as it appears...

Found one more nit in for the nginx package; please fix that, and I'm OK with merging this.

tohojo reviewed (approved):
Excellent. Thanks for sticking with this! :)

tohojo commented:
(OK to merge once the CI run has finished)

schm0 commented:
Is it possible that there is a typo in the hook.sh script?
The path to acme.sh seems wrong?
ACME=/usr/lib/acme/acme.sh
but the actual path is /usr/lib/acme/client/acme.sh

hgl commented:
Thanks, PR sent.

stokito commented:
Can anyone add the hotplug hook to `emailrelay`? If it uses a tls_cert then it should be reloaded.
Does anybody know how to do that?
